### ISSUE TITLE
Shorthand for "pattern matching" on an object parameter

### ISSUE THREAD

**Issue by copumpkin** [google/jsonnet#119]
One thing I really miss when writing jsonnet after writing Nix for a while is the ability to deconstruct an object being passed in as a parameter. For example:

```nix
{ a, b, c }: a + b +c
```

Is a one-variable lambda that expects the caller to pass it a dict with those three keys in it. You can also set defaults and match on the aggregate.

Would it be possible to get something similar in jsonnet?

**Comment by sparkprime** [comment]
So you're not really asking to omit the function keyword, right?  You're asking for a way to write the following in half as much space:

```
local func(T)
    local a = T.a, b = T.b, c = T.c;
    a + b + c;
func({ a: 1, b: 2, c: 3 })
```

One thing I've realized recently is that people want to use functions a lot more than expected.  And in the design they are pretty rudimentary (compared to other languages), with not much sugar to encourage conciseness.

I did for a while flirt with the idea of allowing func{ a: 1, b: 2, c: 3 } as in Lua.  But in Jsonnet that makes it quite hard to see whether func is indeed a func or in fact an object being extended.  So I think the parens have to stay.

One option is to add something like Python named variables and kwargs

```
local func(a, b, c)
    a + b + c;
func(a=1, b=2, c=3)
```
```
func(**{ a: 1, b: 2, c: 3 })
```

"Borrow from Python" has been a good default in other areas.

**Comment by sparkprime** [comment]
@oconnorr as the word "nix" was mentioned

**Comment by oconnorr** [comment]
I think pattern matching is a great thing, but there are some design choices to be made.

Nix has a moderately complex syntax for pattern matching on tuples

{ foo({ a, b }) = a + b }

Does foo require an object whose set of fields is exactly a and b, or does it always slice an accept any object with at least a and b fields?  Nix supports both with special syntax for the second case.

Is the binding lazy so that even if the parameter doesn't have one of the fields, if that field is never used in the computation then maybe that is okay?

Nix also has this ? syntax in the pattern which fills in fields with default values in case the parameter is missing the named field.

**Comment by copumpkin** [comment]
@sparkprime yeah, not looking to omit the fairly heavyweight function keyword (although I would love that too :smile:). I'm fine with the python splat operator, but with the considerations @oconnorr raises. The only thing I could see us losing from treating it as a splat, rather than an explicit dict, is the sort of thing we do fairly often which is to peel a couple of arguments (possibly with defaults) out of a dict being passed in, but also passing the dict through wholesale to another call.

**Comment by sparkprime** [comment]
How do you do that (peel off and pass the whole dict through) in Nix?

**Comment by copumpkin** [comment]
oh, sorry.

foo = { bar, baz, ...}@foo: doSomethingWith [ bar baz ] (someOtherFunction foo);

**Comment by benley** [comment]
Relatedly, I would love to be able to do things like this:
```
local (x, y, z) = std.split("foo bar bas", " ")
```

**Comment by sparkprime** [comment]
Is this addressed by #164 ?  I know it's not exactly what you asked for but it covers the obvious use case.

**Comment by copumpkin** [comment]
I love the default arguments you've added, but is there a splat operator today? With splatting I think it's pretty close to what I'm talking about, but probably wouldn't cover @benley's use case.
